Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs

```diff
@@ -0,0 +1,346 @@
+        ref AfterConstructionChangeEntityEvent args)
+    {
+        if (!XformQuery.TryGetComponent(ent.Owner, out var constructXform))
+            return;
+
+        var query = EntityQueryEnumerator<MultipartMachineComponent>();
+        while (query.MoveNext(out var uid, out var machineComp))
+        {
+            var machine = (uid, machineComp);
+            if (!IsMachineInRange(machine, constructXform.LocalPosition))
+                continue; // This part is outside the max range of the machine, ignore
+
+            foreach (var part in machineComp.Parts.Values)
+            {
+                if (args.Graph == part.Graph &&
+                    (args.PreviousNode == part.ExpectedNode || args.CurrentNode == part.ExpectedNode))
+                {
+                    Rescan(machine);
+                    break; // No need to scan the same machine again
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Handles when a machine part entity has been anchored or unanchored by a user.
+    /// We might be able to link an unanchored part to a machine, but anchoring a constructable entity,
+    /// which machine parts are, will require a rescan of all machines within range as we have no idea
+    /// what machine it might be a part of.
+    /// </summary>
+    /// <param name="ent">Machine part entity that has been anchored or unanchored.</param>
+    /// <param name="args">Args for this event, notably the anchor status.</param>
+    private void OnPartAnchorChanged(Entity<MultipartMachinePartComponent> ent, ref AnchorStateChangedEvent args)
+    {
+        if (!args.Anchored)
+        {
+            if (!TryComp<MultipartMachinePartComponent>(ent.Owner, out var part) || !part.Master.HasValue)
+                return; // This is not an entity we care about
+
+            // This is a machine part that is being unanchored, rescan its machine
+            if (!TryComp<MultipartMachineComponent>(part.Master, out var machine))
+                return;
+
+            Rescan((part.Master.Value, machine));
+            return;
+        }
+
+        // We're anchoring some construction, we have no idea which machine this might be for
+        // so we have to just check everyone in range and perform a rescan.
+        if (!XformQuery.TryGetComponent(ent.Owner, out var constructXform))
+            return;
+
+        var query = EntityQueryEnumerator<MultipartMachineComponent>();
+        while (query.MoveNext(out var uid, out var machineComp))
+        {
+            var machine = (uid, machineComp);
+            if (!IsMachineInRange(machine, constructXform.LocalPosition))
+                continue; // This part is outside the max range of the machine, ignore
+
+            if (Rescan(machine) && HasPartEntity(machine, ent.Owner))
+                return; // This machine is using this entity so we don't need to go any further
+        }
+    }
```

Comment:
Use EntityLookup as it will vary between entityquery vs range checks as required.